### ISSUE TITLE
Reflectable builder fails when using types from dart:ui

### ISSUE THREAD

**Issue by fertrig** [dart-lang/reflectable#172]
When using `dart:ui` types as constructor parameter types of classes that are reflected on, the builder fails with message:

```
NoSuchMethodError: The getter 'library' was called on null.
Receiver: null
Tried calling: library
```

I published a flutter repo with test files that demonstrate the behavior. The tests are [here](https://github.com/fertrig/flutter_test_reflectable/tree/master/test). There are 3 tests:

1. `dart_ui_color_test.dart` uses the `Color` type from `dart:ui`. The builder _fails_ on this file.
2. `dart_ui_text_align.dart` uses the `TextAlign` type from `dart:ui`. The builder also fails on this file.
3. `flutter_services_test.dart` uses the `TextCapitalization` type from `package:flutter/services.dart`. The builder _successfully_ generates code for this file.

I am using this command to run the reflectable builder:

```
flutter packages pub run build_runner build test
```

From looking at the dart:ui source code I can see that the types that fail are inside of part-of files. When I import the part-of files directly in the test files, the builder works. However, importing part-of files is a compile-time error, which the builder seems to ignore.

It may also be that the issue is related to dart:ui since it is part of the sdk.

I'd expect that importing dart:ui should be sufficient to successfully generate reflectable code using the types dart:ui exposes.

**Comment by eernstg** [comment]
Debugging: The analyzer data on the `staticElement` of an expression turns out to be null, which is the reason why the subsequent lookup of `library` fails. This could indicate that the given expression has an error, or the library hasn't been resolved, and those things shouldn't happen here. No solution yet.

**Comment by eernstg** [comment]
There is a new branch `survive_172_may19`. It is unfortunately only half a work-around. I'm traveling next week and won't be able to do much for a week, but that branch `survive_172_may19` at least introduces some changes that allows code generation to survive the unexpected null from `staticElement`.

This is not a complete solution at all, because the generated code will then contain expressions where the addition of imports and import prefixes was omitted. For example, we will generate expressions like the following (one expression per line):

```
TextAlign.start
const Color.fromARGB(255, 50, 50, 50)
```

But these expressions are compile-time errors, because "everything" in the generated library will otherwise use library prefixes (because the generated code may import arbitrary libraries among the ones that the "program" uses, so there could be name clashes that don't otherwise exist in the program if we were to import them all without a prefix).

It is likely that these generated files will work if the following is added:

```dart
// ignore:unused_import
import "dart:ui";
```

This is of course not an acceptable solution, but it might be useful for some experiments at this point, and then I'll look into the underlying null issue when I come back.

**Comment by fertrig** [comment]
I tried the branch `survive_172_may19` on the test repo I posted before. The code generation fails in a different way:

```
[SEVERE] reflectable:reflectable on test/dart_ui_color_test.dart (cached):
Cannot obtain analysis result for expression const Color.fromARGB(255, 50, 50, 50). Required imports and library prefixes may be missing.

[SEVERE] reflectable:reflectable on test/dart_ui_text_align_test.dart (cached):
Cannot obtain analysis result for expression TextAlign.start. Required imports and library prefixes may be missing.
```

These errors seem related to what you had described. I wasn't sure where to place the snippet you suggested:
```
// ignore:unused_import
import "dart:ui";
```
I placed it on the test file and on the source file. It failed in the same way in both cases.

**Comment by eernstg** [comment]
Right, `survive_172_may19` allows code generation to complete, but it will complain in exactly the way you described (`[SEVERE] .. imports and library prefixes may be missing`) when it encounters the situation where `staticElement` on an expression returns null (that is, the cases that your code demonstrated).

However, in the case where that problem arises because something from a part of 'dart:ui' is being referenced, it should be enough to add `import 'dart:ui';` to the _generated_ file.

With that, generated expressions like `TextAlign.start` will no longer be a compile-time error, they will refer to the declarations in 'dart:ui' as intended.

That's the reason why I called it "half a work-around": It is necessary to change the generated file by adding that import. This is no real solution, of course, and you may not want to use it at all. I just thought I would mention it, and then we'll have an actual solution to the problem later.

**Comment by fertrig** [comment]
I see. Adding `import 'dart:ui';` to the generated file does work. Thanks.

I realize the fix is partial and not fully tested. However, I found an issue with this partial fix and the capability `typeAnnotationQuantifyCapability`. For what it's worth, here is the issue:

When `typeAnnotationQuantifyCapability` is used, this partial fix doesn't correctly generate the reflected types for parameter mirrors. The reflected types appear as `dynamic` not as their actual types.

The reflected type of a parameter that uses types from dart:ui is not reflected on:
```
// FAILS: expected Color, actual dynamic
expect(constructorMirror.parameters[0].type.reflectedType, Color);
```
Here is a test case showing the issue: https://github.com/fertrig/flutter_test_reflectable/blob/master/test/dart_ui_color_test.dart

**Comment by cowboyd** [comment]
I think I'm seeing the same or similar issue when trying to generate reflection capabilities for Flutter classes, specifically  `material.MaterialApp`

Things like `widgets.Text` and `widgets.Title` work just fine, but whenever I try to generate reflection capabilities for anything in the `"package:flutter/material.dart"` library I get this error:

```
$ flutter packages pub run build_runner build --verbose
[INFO] Generating build script...
[INFO] Generating build script completed, took 300ms

[INFO] BuildDefinition:Initializing inputs
[INFO] BuildDefinition:Reading cached asset graph...
[INFO] BuildDefinition:Reading cached asset graph completed, took 111ms

[INFO] BuildDefinition:Checking for updates since last build...
[INFO] BuildDefinition:Checking for updates since last build completed, took 581ms

[INFO] Build:Running build...
[INFO] Build:Running build completed, took 13ms

[INFO] Build:Caching finalized dependency graph...
[INFO] Build:Caching finalized dependency graph completed, took 43ms

[SEVERE] reflectable:reflectable on lib/reflector.dart (cached):

NoSuchMethodError: The getter 'isDartCore' was called on null.
Receiver: null
Tried calling: isDartCore
dart:core                                                               Object.noSuchMethod
package:reflectable/src/builder_implementation.dart 3088:17             _ImportCollector._getPrefix
package:reflectable/src/builder_implementation.dart 4266:37             _extractConstantCode.typeAnnotationHelper
package:reflectable/src/builder_implementation.dart 4282:33             _extractConstantCode.helper
package:reflectable/src/builder_implementation.dart 4408:16             _extractConstantCode
package:reflectable/src/builder_implementation.dart 2213:16             _ReflectorDomain._extractDefaultValueCode
package:reflectable/src/builder_implementation.dart 704:21              _ReflectorDomain._constructorCode.<fn>
dart:_internal                                                          ListIterable.join
package:reflectable/src/builder_implementation.dart 708:8               _ReflectorDomain._constructorCode
package:reflectable/src/builder_implementation.dart 1363:23             _ReflectorDomain._classMirrorCode.<fn>
dart:core                                                               Iterable.join
package:reflectable/src/builder_implementation.dart 4254:50             _formatAsMap
package:reflectable/src/builder_implementation.dart 1357:26             _ReflectorDomain._classMirrorCode
package:reflectable/src/builder_implementation.dart 970:13              _ReflectorDomain._generateCode.<fn>.<fn>
dart:core                                                               Iterable.join
package:reflectable/src/builder_implementation.dart 4247:28             _formatAsList
package:reflectable/src/builder_implementation.dart 994:31              _ReflectorDomain._generateCode
package:reflectable/src/builder_implementation.dart 141:21              ReflectionWorld.generateCode.<fn>
dart:_internal                                                          ListIterable.join
package:reflectable/src/builder_implementation.dart 4254:50             _formatAsMap
package:reflectable/src/builder_implementation.dart 156:12              ReflectionWorld.generateCode
package:reflectable/src/builder_implementation.dart 3956:25             BuilderImplementation._generateNewEntryPoint
package:reflectable/src/builder_implementation.dart 4056:35             BuilderImplementation.buildMirrorLibrary
package:reflectable/reflectable_builder.dart 29:37                      ReflectableBuilder.build
package:build                                                           runBuilder
package:build_runner_core/src/generate/build_impl.dart 460:19           _SingleBuild._runForInput.<fn>.<fn>.<fn>
package:build_runner_core/src/generate/performance_tracker.dart 303:15  _NoOpBuilderActionTracker.trackStage
package:build_runner_core/src/generate/build_impl.dart 458:23           _SingleBuild._runForInput.<fn>.<fn>
package:timing/src/timing.dart 222:44                                   NoOpTimeTracker.track
package:build_runner_core/src/generate/build_impl.dart 417:22           _SingleBuild._runForInput.<fn>
package:pool/pool.dart 127:28                                           Pool.withResource
package:build_runner_core/src/generate/build_impl.dart 413:17           _SingleBuild._runForInput
package:build_runner_core/src/generate/build_impl.dart 363:38           _SingleBuild._runBuilder.<fn>
dart:async                                                              Future.wait
package:build_runner_core/src/generate/build_impl.dart 362:36           _SingleBuild._runBuilder
package:build_runner_core/src/generate/build_impl.dart 309:20           _SingleBuild._runPhases.<fn>.<fn>
package:build_runner_core/src/generate/performance_tracker.dart 187:15  _NoOpBuildPerformanceTracker.trackBuildPhase
package:build_runner_core/src/generate/build_impl.dart 305:47           _SingleBuild._runPhases.<fn>
package:timing/src/timing.dart 222:44                                   NoOpTimeTracker.track
package:build_runner_core/src/generate/build_impl.dart 299:32           _SingleBuild._runPhases
package:build_runner_core/src/logging/logging.dart 25:30                logTimedAsync
package:build_runner_core/src/generate/build_impl.dart 260:26           _SingleBuild._safeBuild.<fn>
dart:async                                                              runZoned
package:build_runner_core/src/generate/build_impl.dart 255:5            _SingleBuild._safeBuild
package:build_runner_core/src/generate/build_impl.dart 205:24           _SingleBuild.run
package:build_runner_core/src/generate/build_impl.dart 93:42            BuildImpl.run
package:build_runner_core/src/generate/build_runner.dart 24:14          BuildRunner.run
package:build_runner                                                    BuildCommand.run
package:args/command_runner.dart 197:27                                 CommandRunner.runCommand
package:args/command_runner.dart 112:25                                 CommandRunner.run.<fn>
dart:async                                                              new Future.sync
package:args/command_runner.dart 112:14                                 CommandRunner.run
package:build_runner                                                    run
.dart_tool/build/entrypoint/build.dart 23:22                            main

[SEVERE] Build:
Failed after 115ms
```

Here is the commit that triggers the error https://github.com/cowboyd/flutterscript/commit/80a0b0e0e515e940fd2e82d581b71bb2785883f2 

To reproduce:

```
$ git clone https://github.com/cowboyd/flutterscript -b try-to-reflect-material-app
$ flutter packages get
$ flutter packages pub run build_runner build 
```

**Comment by eernstg** [comment]
This is the same issue as this one: https://github.com/dart-lang/build/issues/733, so we're blocked on that.

**Comment by eernstg** [comment]
For [this issue](https://github.com/dart-lang/reflectable/issues/172#issuecomment-501547744), checking that this is the same issue:

The `LibraryElement` parameter which is null when `isDartCore` is invoked on it is obtained by evaluating  `typeName.type.element.library` where `typeName` is a `TypeAnnotation`. The `typeName` turns out to be `Locale`, which is defined in `dart:ui`; it could not be resolved: its `type` is the `UndefinedTypeImpl`, whose `library` is null.

**Comment by fertrig** [comment]
@eernstg, this issue has been blocked on [dart-lang/build#733](https://github.com/dart-lang/build/issues/733), which is now resolved. Do you know if reflecting on dart:ui types is still an issue? I'll try the latest version of reflectable in the next few days.

**Comment by eernstg** [comment]
It's surely still an issue because I haven't had a chance to look into that solution and use it in reflectable. I hope the underlying problem has been resolved and reflectable can be adjusted to do what's needed, but it's not going to be possible for me to look into it before the upcoming vacation. But there's hope now! ;-)

**Comment by natebosch** [comment]
In theory, nothing should need to happen in this package - someone will need to try it and confirm.

You'll want to start with a `pub upgrade`. If you get `build_resolvers` version `1.2.0` or higher, things should work.

**Comment by eernstg** [comment]
Fantastic! In that case it should just be working:

```
  build_resolvers:
    ...
    version: "1.2.1"
  ...
  reflectable:
    ...
    version: "2.2.0"
```

**Comment by eernstg** [comment]
Ran a small test: It is actually possible at this point to generate stuff like the following:

```
import 'dart:ui' as prefix1;
...

final _data = <r.Reflectable, r.ReflectorData>{
  ...
        r.NonGenericClassMirrorImpl(
          r"Color",
          r"dart.ui.Color",
          7,
          2,
          const prefix0.MyReflectable(),
          const <int>[-1],
          null,
          null,
          -1,
          {
            r"lerp": () => prefix1.Color.lerp,
            r"alphaBlend": () => prefix1.Color.alphaBlend
          },
          ...
        },
  ...
};
```

So I think I'll close this issue as resolved, effortlessly (maybe we should create a new label for that ;-). Please open new ones if something is broken after all.

**Comment by fertrig** [comment]
@eernstg, how did you get it to reflect on `dart.ui.Color`? I have tried many different combinations unsuccessfully.

On [this test file](https://github.com/fertrig/flutter_test_reflectable/blob/testing-221/test/class_with_dart_ui_fields_test.dart), I attempt to reflect on a Color field. Its reflected type returns as dynamic. I was expecting its reflected type to be Color.

And on [this file](https://github.com/fertrig/flutter_test_reflectable/blob/testing-221/test/dart_ui_global_quantify.dart), I attempt to reflect on the Color type using the `GlobalQuantifyCapability` annotation. However, the reflector does not match anything. I tried many different combinations of the class name pattern.

I can open a new issue. However, I would like to make sure I have the correct setup since I couldn't get it to generate any class mirror for `dart.ui.Color`.

**Comment by eernstg** [comment]
Of course, I should have saved the code for that. ;-)

However, I recreated the situation where the tiny example in https://github.com/eernstg/reflectable-flutter is using the `Color` class from 'dart:ui' and updated that repo. So you should be able to use that.

I think the issue is that the analyzer reports the full name of such classes in a different way than expected ('dart:ui' is not a platform library even though it can be imported using a directive of the form `import 'dart:<something>';`). So `r'^dart\.ui\.Color$'` won't match that class, but a plain `\.Color$` does match it. Not sure who would know more about this, but you should still be able to make some progress.

**Comment by fertrig** [comment]
I tried the example. However, I don't see it reflecting on the Color type. I was expecting the reflectable build to generate reflectable code for the Color type. 

It is only generating reflectable code for:
```
@myReflectable
class A {...}
```

It is _not_ generating reflectable code for:
```
@GlobalQuantifyCapability(r'\.Color$', myReflectable)
```
The `GlobalQuantifyCapability` annotation does not seem to match anything.

I am using the latest version of flutter:
```
Flutter 1.12.13+hotfix.5 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 27321ebbad (4 weeks ago) • 2019-12-10 18:15:01 -0800
Engine • revision 2994f7e1e6
Tools • Dart 2.7.0
```

Could you please confirm you see reflectable code generated for the Color type?

**Comment by eernstg** [comment]
Looking at `lib/main.reflectable.dart` from a checkout of test_reflectable, I see the following:

```dart
...
final _data = <r.Reflectable, r.ReflectorData>{
  const prefix0.MyReflectable(): r.ReflectorData(
      <m.TypeMirror>[
        ...,
        r.NonGenericClassMirrorImpl(
            r"Color",
            r"dart.ui.Color",
            7,
            1,
            const prefix0.MyReflectable(),
            const <int>[-1],
            null,
            null,
            -1,
            {
              r"lerp": () => prefix1.Color.lerp,
              r"alphaBlend": () => prefix1.Color.alphaBlend
            },
            {},
            {
              r"": (b) => (value) => b ? prefix1.Color(value) : null,
              r"fromARGB": (bb) => (a, r, g, b) =>
                  bb ? prefix1.Color.fromARGB(a, r, g, b) : null,
              r"fromRGBO": (bb) => (r, g, b, opacity) =>
                  bb ? prefix1.Color.fromRGBO(r, g, b, opacity) : null
            },
            -1,
            -1,
            const <int>[-1],
            null,
            {
              r"==": 1,
              r"toString": 0,
              r"noSuchMethod": 1,
              r"hashCode": 0,
              r"runtimeType": 0,
              r"withAlpha": 1,
              r"withOpacity": 1,
              r"withRed": 1,
              r"withGreen": 1,
              r"withBlue": 1,
              r"computeLuminance": 0,
              r"value": 0,
              r"alpha": 0,
              r"opacity": 0,
              r"red": 0,
              r"green": 0,
              r"blue": 0,
              r"lerp": 4,
              r"alphaBlend": 3
            })
      ],
      ...
  )
};
```

Repeating the steps that created the above, I hit the error `Too many positional arguments` which is caused by the issue reported as #194. So I can confirm that it worked on Jan 6th. I'll take another look at this when I've fixed #194.

**Comment by eernstg** [comment]
Reflectable 2.1.1+1 has been published; it requires the upper bound '0.39.3' for the analyzer version. A more lasting solution to #194 is upcoming.

**Comment by eernstg** [comment]
@fertrig, are you still not seeing any code generated for 'dart:ui' declarations?

**Comment by fertrig** [comment]
I still do not see any code generated for 'dart:ui' declarations.

I cloned the [example repo](https://github.com/eernstg/reflectable-flutter) again. I set these dependencies in pubspec.yaml:

```
  reflectable: 2.2.1+1
  analyzer: '<= 0.39.3'
```

It still doesn't generate reflectable code for 'dart:ui'. 

@eernstg, let me know if it works for you.

**Comment by natebosch** [comment]
The version of `flutter` matters here. In some older versions of flutter we don't detect the fact that we are running with a flutter SDK and don't add it's summary. cc @jakemac53

**Comment by fertrig** [comment]
Here's the flutter version I used:
```
Flutter 1.12.13+hotfix.5 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 27321ebbad (6 weeks ago) • 2019-12-10 18:15:01 -0800
Engine • revision 2994f7e1e6
Tools • Dart 2.7.0
```
It is the latest stable release.

**Comment by natebosch** [comment]
I don't think we expect it to work in flutter stable currently. There is an issue somewhere tracking work that could fix it for flutter stable, but I can't find it... @jakemac53 can you link it here?

**Comment by fertrig** [comment]
I see. I tried it with the latest master channel release and it worked. It generated reflectable code for 'dart:ui'. Thanks!

**Comment by jakemac53** [comment]
https://github.com/dart-lang/build/issues/2528 is the related issue

**Comment by jakemac53** [comment]
Note that this should be resolved now for stable and older flutter versions
